torch_geometric/data/graph_store.py

```diff
@@ -92,15 +116,134 @@ def get_edge_index(self, *args, **kwargs) -> EdgeTensorType:
         """
         edge_attr = self._edge_attr_cls.cast(*args, **kwargs)
         edge_attr.layout = EdgeLayout(edge_attr.layout)
+        # Override is_sorted for CSC and CSR:
+        edge_attr.is_sorted = edge_attr.is_sorted or (edge_attr.layout in [
+            EdgeLayout.CSC, EdgeLayout.CSR
+        ])
         edge_index = self._get_edge_index(edge_attr)
         if edge_index is None:
             raise KeyError(f"An edge corresponding to '{edge_attr}' was not "
                            f"found")
         return edge_index
 
-    # TODO implement coo(), csc(), csr() methods on GraphStore, which perform
-    # conversions of edge indices between formats. These conversions can also
-    # automatically be performed in `get_edge_index`
+    def _to_layout(self, to_layout: EdgeLayout,
+                   store: bool = False) -> ConversionOutputType:
+        # Obtain all edge attributes, grouped by type:
+        edge_attrs = self.get_all_edge_attrs()
+        edge_type_to_attrs: Dict[Any, List[EdgeAttr]] = defaultdict(list)
+        for attr in edge_attrs:
+            edge_type_to_attrs[attr.edge_type].append(attr)
+
+        # Convert layouts for each attribute from its most favorable original
+        # layout to the desired layout. Store permutations of edges if
+        # necessary as part of the conversion:
+        row_dict, col_dict, perm_dict = {}, {}, {}
+        for _, edge_attrs in edge_type_to_attrs.items():
+            edge_layouts = [edge_attr.layout for edge_attr in edge_attrs]
+
+            # Ignore if requested layout is already present:
+            if to_layout in edge_layouts:
+                from_attr = edge_attrs[edge_layouts.index(to_layout)]
+                row, col = self.get_edge_index(from_attr)
+                perm = None
+
+            # Convert otherwise:
+            else:
+                # Pick the most favorable layout to convert from. We always
+                # prefer CSR or CSC to COO:
+                from_attr = None
+                if EdgeLayout.CSC in edge_layouts:
+                    from_attr = edge_attrs[edge_layouts.index(EdgeLayout.CSC)]
+                elif EdgeLayout.CSR in edge_layouts:
+                    from_attr = edge_attrs[edge_layouts.index(EdgeLayout.CSR)]
+                else:
+                    from_attr = edge_attrs[edge_layouts.index(EdgeLayout.COO)]
+
+                from_tuple = self.get_edge_index(from_attr)
+
+                # Convert to the new layout:
+                if to_layout == EdgeLayout.COO:
+                    if from_attr.layout == EdgeLayout.CSR:
+                        adj = SparseTensor(
```

Comment:
Can we drop `SparseTensor` here? We can just convert indices back to COO, and, use `sort_edge_index`.

Reply:
Not sure why we would need `sort_edge_index` here?